### ISSUE TITLE
Newton solver crashes due to uninitialised PETScVector/PETScMatrix

### ISSUE THREAD

**Issue by chrisrichardson** [FEniCS/dolfinx#214]
Trying to use the `NewtonSolver` in Python, it needs to have A, P and b set to valid PETSc objects
before running. There is a callback to `NonlinearProblem::form()` which should allow this setup to take place, however there doesn't seem to be a way to reset the PETSc `Vec` and `Mat` objects to point to new values. e.g.

```
from dolfin import *
from dolfin.fem.assemble import assemble
import dolfin.la as la

# Class for interfacing with the Newton solver                                                                                   
class NLEquation(NonlinearProblem):
    def __init__(self, a, L):
        NonlinearProblem.__init__(self)
        self.L = L
        self.a = a

    def F(self, b, x):
        print(b.size())
        assemble(b, self.L)
    def J(self, A, x):
        assemble(A, self.a)

    def form(self, A, P, b, x):
        # How to reassign Vec and Mat objects here?
        b = assemble(self.L)
        A = assemble(self.a)
        P = assemble(self.a)
        print(b.size())

# Create mesh and build function space                                                                                           
mesh = UnitSquareMesh(MPI.comm_world, 16, 16)
P1 = FiniteElement("Lagrange", mesh.ufl_cell(), 1)
Q = FunctionSpace(mesh, P1)

du = TrialFunction(Q)
v  = TestFunction(Q)

# Define functions                                                                                                               
u = Function(Q)  # current solution                                                                                              

L = u*u*v*dx
a = derivative(L, u, du)

problem = NLEquation(a, L)
solver = NewtonSolver(MPI.comm_world)
solver.parameters["relative_tolerance"] = 1e-6

solver.solve(problem, u.vector())
```

**Comment by garth-wells** [comment]
Use `assemble(b, self.L)` if `b` exists, otherwise 
```
b = PETScVector(self.L.function_space(0).dofmap().index_map())
```

Also, the Newton methods should return something.

**Comment by chrisrichardson** [comment]
I don't think this will work. The problem is that NewtonSolver holds a NULL Petsc Vec at instantiation and I can't see how to reassign it.

**Comment by garth-wells** [comment]
It shouldn't hold an empty vector - it should be a pointer.

**Comment by chrisrichardson** [comment]
It is a `std::shared_ptr<la::PETScVector>` - but it is initialised with `new la::PETScVector()` and I can't see how to reset it to something else...

**Comment by garth-wells** [comment]
Fixed by #223.
